src/main/cpp/RobotContainer.cpp

```diff
@@ -4,18 +4,131 @@
+#include <frc/smartdashboard/SmartDashboard.h>
 
-  // Configure the button bindings
-  ConfigureButtonBindings();
+RobotContainer::RobotContainer()
+{
+    // Initialize the hardware.
+    VOKC_CALL(this->InitHardware());
+
+    // Initialize subsystems.
+    VOKC_CALL(this->InitDrivetrain());
+
+    // Configure the button bindings
+    ConfigureButtonBindings();
+}
+
+void RobotContainer::ConfigureButtonBindings()
+{
+    // Configure your button bindings here
+}
+
+frc2::Command *RobotContainer::GetAutonomousCommand()
+{
+    // An example command will be run in autonomous
+    return &m_autonomousCommand;
+}
+
+bool RobotContainer::InitHardware()
+{
+    // Initialize the hardware interface.
+    hardware_ = std::make_unique<HardwareInterface>();
+
+    // Initialize sub-hardware interfaces.
+    hardware_->actuators = std::make_unique<ActuatorInterface>();
+    hardware_->sensors = std::make_unique<SensorInterface>();
+
+    // Initialize the actuators.
+    OKC_CALL(this->InitActuators(hardware_->actuators.get()));
+
+    // Set up sensors.
+    OKC_CALL(
+        this->InitSensors(*hardware_->actuators, hardware_->sensors.get()));
+
+    return true;
+}
+
+bool RobotContainer::InitActuators(ActuatorInterface *actuators_interface)
+{
+    OKC_CHECK(actuators_interface != nullptr);
+
+    // Initialize drivetrain motors.
+    actuators_interface->left_motor_1 =
+        std::make_unique<rev::CANSparkMax>(LEFT_MOTOR_1, BRUSHLESS);
+    actuators_interface->left_motor_2 =
+        std::make_unique<rev::CANSparkMax>(LEFT_MOTOR_2, BRUSHLESS);
+    actuators_interface->left_motor_3 =
+        std::make_unique<rev::CANSparkMax>(LEFT_MOTOR_3, BRUSHLESS);
+    actuators_interface->right_motor_1 =
+        std::make_unique<rev::CANSparkMax>(RIGHT_MOTOR_1, BRUSHLESS);
+    actuators_interface->right_motor_2 =
+        std::make_unique<rev::CANSparkMax>(RIGHT_MOTOR_2, BRUSHLESS);
+    actuators_interface->right_motor_3 =
+        std::make_unique<rev::CANSparkMax>(RIGHT_MOTOR_3, BRUSHLESS);
+
+    return true;
 }
 
-void RobotContainer::ConfigureButtonBindings() {
-  // Configure your button bindings here
+bool RobotContainer::InitSensors(const ActuatorInterface &actuators,
+                                 SensorInterface *sensor_interface)
+{
+    OKC_CHECK(sensor_interface != nullptr);
+
+    // Initialize navX.
+    try
+    {
+        sensor_interface->ahrs = std::make_unique<AHRS>(frc::SPI::Port::kMXP);
+    }
+    catch (std::exception &ex)
+    {
+        std::string what_string = ex.what();
+        std::string err_msg("Error instantiating navX MXP:  " + what_string);
+        const char *p_err_msg = err_msg.c_str();
+
+        // Print the error message.
+        OKC_CHECK_MSG(false, p_err_msg);
+    }
+
+    return true;
 }
 
-frc2::Command* RobotContainer::GetAutonomousCommand() {
-  // An example command will be run in autonomous
-  return &m_autonomousCommand;
+bool RobotContainer::InitDrivetrain()
```

Comment:
wait why is  `InitDrivetrain` a thing? more specifically why is it a method of RobotContainer? What calls it? Is this necessary for making the code work in C++ or is this extra stuff that we wouldn't have if we just straight Java->CPP'd it? So we're going for a translate+rewrite instead of a straight translate?

Reply:
I wanted to try this style of setting things up in C++, so it's a bit of a re-write. But once it's established then it should be easy to do for the other subsystems

This is a really bad function name. It should probably be `SetupDrivetrainInterface` and then we'd call drivetrain's init function. I'll fix this in my next PR